### ISSUE TITLE
NegotiateAuthClient.send chokes on valid server responses

### ISSUE THREAD

**Issue by jbash** [marcoesposito1988/http_auth#8]
Given (legal) headers like

```
HTTP/1.1 401 Unauthorized
Date: Thu, 15 Oct 2020 15:19:50 GMT
Server: Apache/2.4.46 (Fedora) OpenSSL/1.1.1g mod_auth_gssapi/1.6.1 mod_wsgi/4.6.8 Python/3.8
WWW-Authenticate: Negotiate
WWW-Authenticate: Basic realm="Keepass DAV data"
Content-Length: 381
Content-Type: text/html; charset=iso-8859-1
```

the "send" method of NegotiateAuthClient will throw an error like `Unsupported authenticate scheme negotiate,basic`. I'm having that problem right now with AuthPass.

### Cause

When it gets a 401 response, that method retrieves the contents of the "WWW-Authenticate" header(s)  from response.headers, which appears to give the comma-separated concatenation of all such headers. It then uses a space as a delimiter to find the first authentication header in that string.

However, Apache, for example, may return *more than one* "WWW-Authenticate" header. A server is allowed to do that. In that case, the string from reponse.headers can be something like `firstscheme,secondscheme param1=foo, param2=bar,param3=baz,thirdscheme`. Note the lack of any whitespace around the first comma. This is the common case if the first scheme is "Basic".

It would also be legal for the server to use a bare comma as a delimiter inside a single header; the whitespace isn't a required element. It's also legal for the server to send tabs instead of spaces, or tabs intermixed with spaces. Null challenges are also legal; a server is allowed to send you a header like `WWW-Authenticate: ,   ,,,,  Basic,, Foo param=bar, Baz`.

See https://tools.ietf.org/html/rfc7235

### Suggested solution

If you just always want to accept the first scheme the server lists, you can probably get away with stripping all leading spaces, tabs, and commas, and then taking the first remaining space, tab or comma as a delimiter.

It's not obvious to me that you *should* always take the first scheme, though. If the server offers "Basic, Digest", you want Digest, not Basic. RFC 7235 gives a formal grammar for the header, and the *right* thing to do is really to be prepared to parse the entire grammar (and discard anything that doesn't parse).

I don't offer any code because I don't know any Dart.

### Random pontification

Obviously the syntax of the WWW-Authenticate header is horrible. Unfortunately, HTTP is a protocol that "just grew", and a lot of it was defined by people who frankly didn't have a clue what they were doing. We're all stuck with it now. Section 4.1 of RFC 7235 specifically talks about the need to be careful when parsing of the WWW-Authenticate header. 

By the way, the word "negotiate" is actually a valid scheme to specify on a "WWW-Authenticate" header line, and it actually gets used in practice. It invokes GSSAPI, which is used for Kerberos and NTLM authentication of HTTP. Having this class named NegotiateAuthClient could easily be confusing.

**Comment by marcoesposito1988** [comment]
Thank you so much for the detailed and constructive issue description! I was not aware of this (yet another sad) corner of the Web.

A proper solution would probably have to wait for [this issue](https://github.com/dart-lang/http/issues/24) to be fixed, but I will try to find a non-fragile workaround for the most common cases. Suggestions (and of course PRs) are always welcome.

My first shot will be:
- split wherever possible (`' '`,`','`, etc), look for the `basic` and `digest` words 
- both present: choose digest since it is more secure
- one present: take it
- neither present: since we don't support more esoteric methods, dump the string and beg the user for forgiveness or a PR

**Comment by marcoesposito1988** [comment]
The fix in #9 was confirmed to work
